drivers: ieee802154: dw1000: implement set_txpower()

loopsmark opened the pull request:
Implementation for setting the transmit power on Qorvo (Decawave) DW1001

This allows the user to increase or decrease the transmit power relative to the default values which are based on a typical IC and an assumed IC to antenna loss of 1.5 dB with a 0 dBi antenna (User Manual - Chapter 7.2.31.4).

Setting TX power below max allowed by regulation allows applications to dynamically reduce range and thereby interference. Plus, as we don't support calibration yet, this can also be used to calibrate TX power level per design/board or per unit for regulatory reasons.

fgrandel commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,12 +769,66 @@ static int dwt_filter(const struct device *dev,
-	LOG_INF("set_txpower not supported %p", ctx);
-
+	if (sys_cfg & DWT_SYS_CFG_DIS_STXP) {
+		/* SmartPower ON */
```
Smart TX Power vs. Manual TX Power (rather than SmartPower ON/OFF) - this is how the manual refers to it

loopsmark replied:
Fixed

fgrandel commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,12 +769,66 @@ static int dwt_filter(const struct device *dev,
+		tx_power |=  ((uint32_t)boostp125 << DWT_TX_POWER_BOOSTP125_SHIFT);
+	} else {
+		/* SmartPower OFF */
+		uint8_t tx_pow_phr = dwt_encode_power_tx(dbm + DWT_TX_POWER_PHR_MOD);
```
I'd remove this redundant extra calculation in this PR as you're setting PHR_MOD to zero anyway. You can introduce a way to reduce PHR power in a later PR. If you can like you can just add another TODO comment.

loopsmark replied:
Outdated

fgrandel commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,12 +769,66 @@ static int dwt_filter(const struct device *dev,
+	} else {
+		/* SmartPower OFF */
+		uint8_t tx_pow_phr = dwt_encode_power_tx(dbm + DWT_TX_POWER_PHR_MOD);
+		tx_power |= ((uint32_t)dwt_encode_power_tx(dbm) << DWT_TX_POWER_TXPOWSD_MASK);
```
_MASK -> _SHIFT ?

loopsmark replied:
Fixed

fgrandel reviewed (changes requested):
Hi @loopsmark, thank you very much for this contribution. As you expected there are a few things that need tweaking. But it's absolutely doable.

As the changes requested are rather involved we'll need further review rounds, so expect me to request further changes even after you've applied all changes requested in this review round.

loopsmark commented on samples/net/sockets/echo_client/prj.conf:
```diff
@@ -1,9 +1,9 @@
 # Generic networking options
 CONFIG_NETWORKING=y
 CONFIG_NET_UDP=y
-CONFIG_NET_TCP=y
```
Done

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,12 +769,66 @@ static int dwt_filter(const struct device *dev,
 }
 
+// From 7.2.31.1 from User Manual
+#define DWT_TX_POWER_CTRL_OFF 7
```
Done

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,12 +769,66 @@ static int dwt_filter(const struct device *dev,
+#define DWT_TX_POWER_TXPOWPHR_SHIFT 8
+#define DWT_TX_POWER_TXPOWSD_SHIFT 16
+
+// These should go into DT probably
```
Done

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,12 +769,66 @@ static int dwt_filter(const struct device *dev,
 	return -ENOTSUP;
 }
 
+// From 7.2.31.1 from User Manual
```
Done

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,12 +769,66 @@ static int dwt_filter(const struct device *dev,
+
+// These should go into DT probably
+#define DWT_TX_POWER_BOOSTP500_MOD 3
+#define DWT_TX_POWER_BOOSTP250_MOD 6
```
Fixed

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,12 +769,66 @@ static int dwt_filter(const struct device *dev,
+    return result;
+}
+
 static int dwt_set_power(const struct device *dev, int16_t dbm)
```
Fixed

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,12 +769,66 @@ static int dwt_filter(const struct device *dev,
+		LOG_ERR("Power %u is not available, power must be below 30", dbm);
+		return -EINVAL;
+	}
 	struct dwt_context *ctx = dev->data;
```
Fixed

fgrandel commented:
After having looked through the [Pixel 6 UWB code](https://android.googlesource.com/kernel/google-modules/uwb/+/refs/heads/android-gs-bluejay-5.10-android12L-d2), manuals, application notes and ETSI UWB regulation (FCC seems to be similar) here come my thoughts re TX Power formula for the DW1000/3000 driver.

**Context:**
- UWB TX power works differently from narrowband TX power we know from other PHYs in the subsystem. This concerns both power level representation in hardware registers and legislation (regulatory power limits).
- The difference in representation in hardware lies in the sensitivity of the effectively radiated power level (per MHz) and bandwidth that results from register settings (ie TX power and pulse generator delay) to many external parameters which is not the case for configured TX power in existing narrowband hardware. The qorvo user manual and application notes mention at least channel (center frequency + bandwidth), PRF, board- and antenna-specific path loss, frame section (STS, SHR, PHR, data), production-specific per-chip gain variations, crystal drift, voltage and temperature. Some of these have non-zero cross-correlation. The effectively radiated power and bandwidth are a function of these parameters and the configured register values.
- The difference in legislation lies in the fact that rather than a single peak radiation value (as for narrowband/DSSS PHYs), a spectral power level mask over the full bandwidth of a channel has been defined that cannot be exceeded anywhere.
- Short frames (<1ms) can be boosted for additional range as the regulation defines mean power per 1ms while peak power may vary within a wider range (within certain limits). Therefore a per-frame boost can be applied for short frames.

**Assumptions:**
- We assume that production-, board- and antenna-specific deviations from nominal power and bandwidth are constant and can be jointly represented as a single calibration pair {nominal_tx_power, nominal_pg_count} per {antenna, channel, PRF} tuple. Setting registers to the calibrated {power/pg delay value} tuple sets the TX power to the maximum power within the allowable spectral mask according to legislation for continuous transmission (frames > 1ms) at nominal environmental temperature and voltage.
- TX power boost values can be directly calculated from frame length (in bytes via symbol period - see APS023 - part 1, section 3).
- Compensating bandwidth for temperature variation is supported in hardware (for both DW1000 and DW3000 - see PG_DELAY calibration based on nominal PG count in the user manual and app note APS023 - part 2, section 4.3).
- Compensating TX power for temperature and voltage is approx. linear with one constant adjustment coefficient per channel (see APS023 - part 1, section 8.3.2 and APS023 - part 2, section 5.3).

If one wants to spare driver API clients the hassle of having to take into account all those parameters and tweaks we need to come up with a driver-internal formula so that clients can simply set a single dB value that is relative to a well-defined zero-dB effective mean power radiation per MHz reference level over the full bandwidth. The driver then takes the responsibility to keep the effective spectral power radiation profile constant across all external parameter changes, i.e. {antenna, channel, PRF, temperature, voltage, frame length}, except for the constant multiplier set via `set_txpower()`.

**I therefore propose that for UWB PHYs calling `set_txpower(0)` SHALL set the effective mean spectral power radiation to the maximum value allowed by regulation over the full bandwidth and that negative/positive values in units of dBr MAY be set which amounts to dividing/multiplying the overall power level by a constant factor over the full bandwidth within the full range allowed by hardware.**

**Proposed changes for this milestone (ie. this PR):**
- We ignore AoA antenna pair setup, frame-length and frame-section dependent power boost and temperature/voltage dependency for now - see the [UWB roadmap](https://github.com/zephyrproject-rtos/zephyr/issues/63953).
- The additional parameters should be easy to add later on, though. A well-defined migration path should exist - please comment accordingly in PR description.
- As long as we do not take into account board/production/antenna variation, power boost and temperature/voltage variation, we use the recommended initial approx. calibration values as defined in the DW1000 datasheet and already present in the current driver (except for one typo which @loopmark discovered and which is to be corrected in this PR) as the 0dB reference point.
- The user may set positive and negative relative dB values (ie. dBr values) via `ieee802154_radio->set_txpower()` to set TX power below or above the nominal calibration value - within the full ranges allowed by the TX_POWER register. Thereby the user can reduce interference in dense networks with short ranges or ensure optimal range while maintaining regulatory compliance under practical conditions within well-defined environmental operating ranges.

**Not in this PR but marked as extension for future milestones:**
- Calculate TX power boost relative to the calibration value on a per-frame basis (either in hardware for DW1000 or in software for DW3000+).
- Configure constant <channel, PRF> calibration values (<antenna, channel, PRF> as soon as AoA/PDoA is to be supported) on a per device level - maybe via calibration values stored in non-volatile memory and configured via driver-specific custom net_mgmt calls. Whenever the user changes the channel or re-configures the PRF value or when the antenna setup is switched at runtime, the calibration values to be used as base values for `set_txpower()`, temperature/voltage compensation and per-frame/frame-section boost SHALL be updated accordingly.
- Configure/calibrate multiple antennas (e.g. up to four) - so that AoA may be accounted for.
- Compensate TX power and pulse generator delay for temperature and voltage changes - either at regular intervals or whenever the measured temperature or power deviates by a given margin from the last compensation level.
- Set per frame-section specific reference values.

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,12 +769,66 @@ static int dwt_filter(const struct device *dev,
+#define DWT_TX_POWER_BOOSTP125_MOD 9
+#define DWT_TX_POWER_PHR_MOD 0
+
+uint8_t dwt_encode_power_tx(float power) {
```
Fixed

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,12 +769,66 @@ static int dwt_filter(const struct device *dev,
 static int dwt_set_power(const struct device *dev, int16_t dbm)
 {
+	if (dbm > 30) {
+		LOG_ERR("Power %u is not available, power must be below 30", dbm);
```
Outdated

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,12 +769,66 @@ static int dwt_filter(const struct device *dev,
+uint8_t dwt_encode_power_tx(float power) {
+	uint8_t coarse = DWT_TX_POWER_CTRL_OFF;
+	uint8_t fineTuneVal = 0;
+	if (power >= 0) {
```
In this case I believe an ASSERT would not work because it is considered that a negative power will use the DWT_TX_POWER_CTRL_OFF value since no negative values can be used according to the documentation. 
Should maybe I explicitly write that as a comment?

loopsmark replied:
Fixed

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,12 +769,66 @@ static int dwt_filter(const struct device *dev,
+	uint8_t fineTuneVal = 0;
+	if (power >= 0) {
+		coarse--;
+		while (power >= DWT_TX_POWER_CTRL_COARSE_STEPS && coarse > 0) {
```
Is this comment still valid?

fgrandel replied:
Yes, this should still be valid - see the summary of our research below which largely confirms the assumptions made in this comment.

fgrandel replied:
```js
function calc_tx_power(target_power) {
    coarse_power_step_x2 = 5;
    fine_power_step_x2 = 1;
    max_coarse_power_step = 0b110;
    max_fine_power_step = 0b11111;

    target_power_int = Math.trunc(target_power);
    target_power_int_x2 = target_power_int << 1;

    coarse_power_steps = Math.min(div_c(target_power_int_x2, coarse_power_step_x2), max_coarse_power_step);
    coarse_power_x2 = coarse_power_steps * coarse_power_step_x2;
    fine_power_steps = div_c(target_power_int_x2 - coarse_power_x2, fine_power_step_x2);
    
    coarse_power_reg = max_coarse_power_step - coarse_power_steps;
    fine_power_reg = fine_power_steps;
    power_reg = coarse_power_reg << 5 | fine_power_reg;
    
    calc_power = coarse_power_steps * 2.5 + fine_power_steps * 0.5;
    
    console.log(`coarse pwr stps: ${coarse_power_steps}, fine pwr stps: ${fine_power_steps}, is valid: ${target_power_int === calc_power}, reg: ${power_reg.toString(16)}`);
}
```

Did it in JS so that you can play around with it in your browser console first. Of course this is a prime example of a function that deserves to be unit tested. I just did a few numbers - so no guarantee the algo is 100% correct.

loopsmark replied:
Your solution is definitely better. I've tested it a bit more and I haven't found any problem.
It's been adopted

loopsmark committed:
drivers: ieee802154: dw1000: DIS_STXP configuration

Add the Kconfig option for DIS_STXP, which allows
to disable the Smart TX Power control

Signed-off-by: Marcos Lopez <loopsmark@gmail.com>

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,12 +769,66 @@ static int dwt_filter(const struct device *dev,
 
-	LOG_INF("set_txpower not supported %p", ctx);
-
+	if (sys_cfg & DWT_SYS_CFG_DIS_STXP) {
```
The  commit [1bc34025ff87300a570bea2a2d5194065b8afbfe](https://github.com/zephyrproject-rtos/zephyr/pull/63887/commits/1bc34025ff87300a570bea2a2d5194065b8afbfe) should fix this.

loopsmark committed:
drivers: ieee802154: dw1000: default value for TX_POWER

Update the value using the current DW1000 User Manual and Table 19

Signed-off-by: Marcos Lopez <loopsmark@gmail.com>

fgrandel commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,11 +769,137 @@ static int dwt_filter(const struct device *dev,
 }
 
-static int dwt_set_power(const struct device *dev, int16_t dbm)
+uint8_t dwt_encode_power_tx(int power)
```
The commit introducing the set_txpower implementation should probably be the last. I think the commits re configurability of smart tx power and the calibration value fix are preparatory and this commit should build upon these.

loopsmark replied:
Fixed

fgrandel reviewed (changes requested):
The PR has been improved in a lot of respects but still has some issues, IMO:
- while commits are now nicely divided into small chunks, commit order should be changed (see arguments in the detailed comments)
- the TX power calculation is not based on calibration values as laid out in the concept comment
- the Kconfig variable for smart mode is missing
- the Coding Guideline check fails (misplaced declarations and missing line spacings)
- the PR contains floating point arithmetics
- I've commented on your questions re the open review comments from the first review round so that you should be able to address them now, I hope

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -137,6 +138,7 @@ static struct dwt_context dwt_0_context = {
 
 		.tx_shr_code = 10,
 		.tx_shr_nsync = DWT_PLEN_128,
+		.dis_stxp = false,
```
Is this commit https://github.com/zephyrproject-rtos/zephyr/pull/63887/commits/5cf2b41e84cf47bab2ec040161479c39e593d179 fixing this?

loopsmark replied:
Now I understand. Until we don't implement a way to change it at run time, it doesn't make sense to have it there.
Removed

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -1488,7 +1491,13 @@ static int dwt_configure_rf_phy(const struct device *dev)
 	}
 
-	if (sys_cfg & DWT_SYS_CFG_DIS_STXP) {
+	if (disable_stxp) {
```
Fix

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,11 +769,137 @@ static int dwt_filter(const struct device *dev,
+		float phr = dwt_decode_power_tx((tx_power >> DWT_TX_POWER_TXPOWPHR_SHIFT) & 0xFF);
+		float sd = dwt_decode_power_tx((tx_power >> DWT_TX_POWER_TXPOWSD_SHIFT) & 0xFF);
+		phr += dbm;
+		if (phr < 0) {
```
Fixed

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,11 +769,137 @@ static int dwt_filter(const struct device *dev,
+		tx_power |= ((uint32_t)dwt_encode_power_tx(phr) << DWT_TX_POWER_TXPOWPHR_SHIFT);
+	} else {
+		/* Smart Transmit Power Control */
+		float norm = dwt_decode_power_tx(tx_power & 0xFF);
```
Fixed, all computations are now in mdB so floats are removed

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,11 +769,137 @@ static int dwt_filter(const struct device *dev,
+static int dwt_set_txpower(const struct device *dev, int16_t dbm)
+{
+	uint32_t tx_power = dwt_reg_read_u32(dev, DWT_TX_POWER_ID, 0);
+	uint32_t sys_cfg = dwt_reg_read_u32(dev, DWT_SYS_CFG_ID, 0);
```
All registry readings are removed, reading the context instead.

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,11 +769,137 @@ static int dwt_filter(const struct device *dev,
+		/* Manual Transmit Power Control */
+		float phr = dwt_decode_power_tx((tx_power >> DWT_TX_POWER_TXPOWPHR_SHIFT) & 0xFF);
+		float sd = dwt_decode_power_tx((tx_power >> DWT_TX_POWER_TXPOWSD_SHIFT) & 0xFF);
+		phr += dbm;
```
Yes, this was always clear. The current TX power setting (AFAIK) could only have values from the table 19/20 since. Anyways the approach have changed and now the function reads directly the (decoded) values from the tables depending on the active channel

loopsmark commented on drivers/ieee802154/ieee802154_dw1000_regs.h:
```diff
@@ -687,6 +687,15 @@
 #define DWT_TX_POWER_ID             0x1E
 #define DWT_TX_POWER_LEN            4
+
+/* Unitos of TX Power Control. From User Manual 7.2.31.1 */
```
Fixed

loopsmark commented on drivers/ieee802154/ieee802154_dw1000_regs.h:
```diff
@@ -687,6 +687,15 @@
+
+/* Unitos of TX Power Control. From User Manual 7.2.31.1 */
+#define DWT_TX_POWER_CTRL_OFF          7
+#define DWT_TX_POWER_CTRL_COARSE_MASK  0xE0U
```
Fixed

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,11 +769,137 @@ static int dwt_filter(const struct device *dev,
+			LOG_WRN("Relative TX power would make TXPOWPHR below minimum, TXPOWPHR=0");
+			phr = 0;
+		} else if (phr > DWT_TX_POWER_CTRL_MAX) {
+			LOG_WRN("Relative TX power would make TXPOWPHR avobe maximum, "
```
Outdated

fgrandel reviewed (dismissed):
We're closing in on a final solution I think. :-)

Just a few more low-level comments. Some of them still blocking, though.

Another review round will be required for the final details.

loopsmark committed:
drivers: ieee802154: dw1000: use configuration to reduce size

Use Kconfig value to select which look up table should be
stored and used for TX_POWER

Signed-off-by: Marcos Lopez <loopsmark@gmail.com>

loopsmark committed:
drivers: ieee802154: dw1000: set_txpower implementation

Add implementation to change the register for Transmit Power Control.
The function will set TX power below or above the nominal
calibration value - within the full ranges allowed by the
TX_POWER register.

Signed-off-by: Marcos Lopez <loopsmark@gmail.com>

loopsmark commented on drivers/ieee802154/Kconfig.dw1000:
```diff
@@ -35,4 +35,14 @@ config IEEE802154_DW1000_INIT_PRIO
 	  be ready first (and sometime gpio should be the very first as spi
 	  might need it too). And of course it has to start before the net stack.
 
+config DISABLE_SMART_TX_POWER_CONTROL
```
Changed

loopsmark commented on drivers/ieee802154/Kconfig.dw1000:
```diff
@@ -35,4 +35,14 @@ config IEEE802154_DW1000_INIT_PRIO
+	  short data frames at this rate (and providing that the frame
+	  transmission rate is < 1 frame per millisecond) it is possible to
+	  increase the transmit power and still remain within regulatory power limits
+	  which are typically specified as average power per millisecond.
```
Added

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -771,11 +771,90 @@ static int dwt_filter(const struct device *dev,
+
+	if (ctx->rf_cfg.dis_stxp) {
+		/* Manual Transmit Power Control */
+		uint16_t phr_sd_mdB;
```
Fixed

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -771,11 +771,90 @@ static int dwt_filter(const struct device *dev,
+		uint8_t phr_sd;
+
+		if (ctx->rf_cfg.prf == DWT_PRF_64M) {
+			phr_sd_mdB = dwt_txpwr_stxp1_64_mdB[dwt_ch_to_cfg[chan] * 4];
```
Yes, you are right, two-dimensional is way more intuitive.
But I cannot remove all of them since the I need to keep the values for the different channels since the channels can be changed at run time.

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -1362,13 +1444,15 @@ static int dwt_configure_rf_phy(const struct device *dev)
 	}
 
-	if (sys_cfg & DWT_SYS_CFG_DIS_STXP) {
+	if (disable_stxp) {
```
Done

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -769,11 +771,90 @@ static int dwt_filter(const struct device *dev,
-static int dwt_set_power(const struct device *dev, int16_t dbm)
+uint8_t dwt_encode_power_tx(int16_t power)
+{
+	uint8_t coarse = DWT_TX_POWER_CTRL_OFF;
```
Outdated

loopsmark commented on drivers/ieee802154/ieee802154_dw1000_regs.h:
```diff
@@ -1949,10 +1973,23 @@ const uint32_t dwt_txpwr_stxp0_64[] = {
 	 0x2B4B6B8B,
 	 0x3A5A7A9A,
-	 0x25456585,
+	 0x25466788,
```
You sure? I believe original value before this PR was 0x25456585, and with this PR it's being changed to 0x25466788.
Am I missing something?

jfischer-no replied:
This does not seem to be correct. I see "5 0x0E082848 0x25456585" in Table 19.

loopsmark committed:
drivers: ieee802154: dw1000: PRF configuration

Add Kconfig option to select the PRF configuration.
The Pulse Repetition Frequency can be either 16 MHz
or 64 MHz

Signed-off-by: Marcos Lopez <loopsmark@gmail.com>

loopsmark commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -1271,6 +1352,7 @@ static int dwt_configure_rf_phy(const struct device *dev)
 {
 	struct dwt_context *ctx = dev->data;
 	struct dwt_phy_config *rf_cfg = &ctx->rf_cfg;
+	bool disable_stxp = rf_cfg->dis_stxp;
```
done

jfischer-no commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -1364,14 +1363,14 @@ static int dwt_configure_rf_phy(const struct device *dev)
 	if (IS_ENABLED(CONFIG_ENABLE_MANUAL_TX_POWER_CONTROL)) {
 		sys_cfg |= DWT_SYS_CFG_DIS_STXP;
-		if (rf_cfg->prf == DWT_PRF_64M) {
+		if (IS_ENABLED(CONFIG_PRF_64MHZ)) {
```
You change breaks configuration per instance, what TBH is not supported by the driver completely but the part you change does it. You have to keep rf_cfg->prf and use devicetree property or runtime configuration.

fgrandel replied:
@jfischer-no I agree that iif the driver supported multiple instances then it would have to maintain instance support here. But this is not the case as you say.

Therefore we this is not a clear-cut decision but a trade off: As a user you want to be able to configure all driver-specific parameters in one place. I had explicitly to @loopsmark to introduce a preliminary change that makes the driver instance capable and moves all KConfig variables to DT. But the contributor considered this to be too much work and too far away from his original intent. Therefore, by Zephyr's reviewer rules, I did not insist to add extra requirements to this PR. My conscious decision was that if I have to choose between exposing the user to an inconsistent configuration API versus getting no contribution at all, we rather leave the multi-instance migration to a a future PR by some contributor that actually needs it.

BTW: Similar decisions are regularly being approved in the sub-system, first and foremost in the nRF5 driver. If we ask for DT configuration here, then Nordic should be the first to make their driver multi-instance capable and move all their configuration to DT.

jfischer-no replied:
Multiinstance support is straightforward, but not after this change. As long as it breaks, my NACK stays.


> BTW: Similar decisions are regularly being approved in the sub-system, first and foremost in the nRF5 driver. If we ask for DT configuration here, then Nordic should be the first to make their driver multi-instance capable and move all their configuration to DT.

Perhaps, but there is a nuance: you can have multiple DW1000 transceivers connected to a SoC, but not multiple nRF5 transceivers. Your statement is still a little confusing, what does Nordic have to do with this PR?

fgrandel replied:
@jfischer-no I follow your argument, that multiple instances make sense for external transceivers but not for nRF5 radios. But my argument re providing a consistent configuration interface to application developers and not extending the scope of PRs unless the contributor agrees is still valid.

So what I propose is that we re-introduce `rf_cfg->prf` to maintain a straightforward migration path towards multi-instance capability. But still we make all new parameters configurable via KConfig until someone else steps up and actually wants to make the driver multi-instance capable. We do not force the contributor to move all configuration variables to DT.

jfischer-no replied:
This is fine as long as the drivers are EXPERIMENTAL/UNSTABLE.

jfischer-no commented on drivers/ieee802154/Kconfig.dw1000:
```diff
@@ -49,4 +49,11 @@ config ENABLE_MANUAL_TX_POWER_CONTROL
 	  increase the transmit power and still remain within regulatory power limits
 	  which are typically specified as average power per millisecond.
 
+config PRF_64MHZ
```
Same as the Kconfig option introduced in the previous patch, but the effect here is more problematic.

fgrandel replied:
See previous comment re Kconfig vs DT, BUT
- please provide this as an option to choose from for improved readability and more importantly the standard provides more than two options and future qorvo chipsets may very well support them.
- All options in this file must be properly prefixed. Please use the prefix introduced for the other options in this file.

loopsmark replied:
Options are now prefixed.
PRF is now a choice.

jfischer-no commented on drivers/ieee802154/ieee802154_dw1000_regs.h:
```diff
@@ -1925,40 +1925,32 @@ const uint8_t dwt_pgdelay_defs[] = {
-	0x0E082848,
-	0x32527292
-};
+#if IS_ENABLED(CONFIG_ENABLE_MANUAL_TX_POWER_CONTROL)
+
+#if IS_ENABLED(CONFIG_PRF_64MHZ)
```
That makes power setting per driver instance impossible. What does it reduce? No flash for additional 24 octets? From the debugger view, how to determine what table is used?

fgrandel replied:
IMO: optional

loopsmark replied:
As proposed:

- PRF reverted to a setting.
- Manual TX Power Control remains a KConfig

jfischer-no commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -767,11 +767,88 @@ static int dwt_filter(const struct device *dev,
+	uint8_t fine_reg;
+	uint8_t reg;
+
+	if (target_mdb < 0) {
+		return 0xC0;
+	} else if (target_mdb > DWT_TX_POWER_CTRL_MAX_MDB) {
+		return 0x1F;
+	}
```
```

	if (target_mdb < 0) {
		return 0xC0;
	}

	if (target_mdb > DWT_TX_POWER_CTRL_MAX_MDB) {
		return 0x1F;
	}

Whatever 0xC0 or 0x1F is...
```

fgrandel replied:
I agree with @jfischer-no that introducing a well-named macro for the magic numbers would be very helpful here.

loopsmark replied:
Fixed

jfischer-no commented on drivers/ieee802154/ieee802154_dw1000_regs.h:
```diff
@@ -687,6 +687,14 @@
 #define DWT_TX_POWER_LEN            4
+
+/* Units of TX Power Control. From User Manual 7.2.31.1 */
+#define DWT_TX_POWER_CTRL_COARSE_MASK          0b110
+#define DWT_TX_POWER_CTRL_COARSE_STEPS_MDB_X2  50
+#define DWT_TX_POWER_CTRL_FINE_MASK            0b11111
```
Do not use binary constants.

fgrandel replied:
...but HEX

loopsmark replied:
Fixed

fgrandel commented on drivers/ieee802154/Kconfig.dw1000:
```diff
@@ -35,4 +35,18 @@ config IEEE802154_DW1000_INIT_PRIO
 
+config ENABLE_MANUAL_TX_POWER_CONTROL
+	bool "Enable Manual TX Power Control"
+	default n
```
Additionally: Please use the correct prefix.

loopsmark replied:
Prefix updated. 
Removed redundant defaults

fgrandel commented on drivers/ieee802154/Kconfig.dw1000:
```diff
@@ -35,4 +35,18 @@ config IEEE802154_DW1000_INIT_PRIO
 	  be ready first (and sometime gpio should be the very first as spi
 	  might need it too). And of course it has to start before the net stack.
 
+config ENABLE_MANUAL_TX_POWER_CONTROL
```
IMO: optional - see argument below

jfischer-no commented on drivers/ieee802154/ieee802154_dw1000_regs.h:
```diff
@@ -1925,40 +1925,29 @@ const uint8_t dwt_pgdelay_defs[] = {
-	0x0E082848,
-	0x32527292
-};
+#ifdef CONFIG_ENABLE_MANUAL_TX_POWER_CONTROL
```
Please do not add any conditions to this header, this will need to be undone later to support multiple instances.

jfischer-no commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -774,11 +774,95 @@ static int dwt_filter(const struct device *dev,
+		dwt_txpwr_mdb = (void *)dwt_txpwr_64_mdb;
+	}
+
+#ifdef CONFIG_IEEE802154_DW1000_ENABLE_MANUAL_TX_POWER_CONTROL
+		/* Manual Transmit Power Control */
+		uint16_t phr_sd_mdb;
+		uint8_t phr_sd;
+
+		phr_sd_mdb = ((uint16_t *)dwt_txpwr_mdb)[dwt_ch_to_cfg[chan]];
+		phr_sd = dwt_encode_power_tx(phr_sd_mdb + mdbm);
+		tx_power |= ((uint32_t)phr_sd << DWT_TX_POWER_TXPOWSD_SHIFT);
+		tx_power |= ((uint32_t)phr_sd << DWT_TX_POWER_TXPOWPHR_SHIFT);
+#else
+		/* Smart Transmit Power Control */
+		uint16_t norm_mdb;
+		uint16_t p500_mdb;
+		uint16_t p250_mdb;
+		uint16_t p125_mdb;
+		uint8_t norm;
+		uint8_t p500;
+		uint8_t p250;
+		uint8_t p125;
+
+		norm_mdb = ((uint16_t (*)[4])dwt_txpwr_mdb)[dwt_ch_to_cfg[chan]][0];
+		p500_mdb = ((uint16_t (*)[4])dwt_txpwr_mdb)[dwt_ch_to_cfg[chan]][1];
+		p250_mdb = ((uint16_t (*)[4])dwt_txpwr_mdb)[dwt_ch_to_cfg[chan]][2];
+		p125_mdb = ((uint16_t (*)[4])dwt_txpwr_mdb)[dwt_ch_to_cfg[chan]][3];
+
+		norm = dwt_encode_power_tx(norm_mdb + mdbm);
+		p500 = dwt_encode_power_tx(p500_mdb + mdbm);
+		p250 = dwt_encode_power_tx(p250_mdb + mdbm);
+		p125 = dwt_encode_power_tx(p125_mdb + mdbm);
+
+		tx_power = norm;
+		tx_power |= ((uint32_t)p500 << DWT_TX_POWER_BOOSTP500_SHIFT);
+		tx_power |= ((uint32_t)p250 << DWT_TX_POWER_BOOSTP250_SHIFT);
+		tx_power |= ((uint32_t)p125 << DWT_TX_POWER_BOOSTP125_SHIFT);
+#endif
```
Appears to be indented too far.
